[MDB IGNORE] ZASR(eforged): A Baystation12 port & modification

Kapu1178 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
### HUGE thanks to TG's LemonInTheDark & Nebula's Lohikar for their knowledge and assistance during this mess.
Ports Baystation's ZAS to TG, replacing Linda in all functionality.
Featuring:
- A highly performant atmospherics solution!
- Explosive decompression!
- Space wind!
- Plasma contamination!
- Destructive temperature mechanics!
- Delta P!
- Lots of Kapu Code that will likely break the laws of thermodynamics!

PROGRESS:

- [X] Basic behavior
- [X] Fires/Hotspots
- [X] Basic atmospheric objects (Canisters, scrubbers)
- [X] Mob-Gas interaction
- [X] XGM gases
- [x] Lavaland atmos
- [x] Advanced atmos machines
- [x] The Supermatter
- [x] Multi-ZAS

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
People really, really like explosive decompression.

## Changelog
:cl:
del: Linda
del: Gas miners
add: Zone Air System
add: Fires now damage windows and walls if they are burning hot enough
add: Objects and people are displaced by pressure differentials
add: Air canisters are now purchasable through Cargo
add: Inflatable barriers and doors for aiding in station repair without venting the whole station
/:cl:

Kapu1178 committed:
Initial commit

Kapu1178 committed:
PAIN AND AGONY

Kapu1178 committed:
Under 1000 errors

Kapu1178 committed:
Checkpoint!

Kapu1178 committed:
Gonna pull the trigger soon

Kapu1178 committed:
Co-authored-by: Gallyus <francinum@users.noreply.github.com>

Kapu1178 committed:
adds immutable mixtures

Kapu1178 committed:
!!!/SIMULATED/ REMOVAL!!!

Kapu1178 committed:
ITS TIME

francinum committed:
ZAS HELL

Kapu1178 committed:
it compiles

francinum committed:
cummies

Kapu1178 committed:
atmostest

Troontjelolo commented:
![zassed](https://user-images.githubusercontent.com/56892466/163835861-1c91c085-1ee0-4af5-838c-66da20fe837b.png)

Sinestia commented:
zas

Kapu1178 committed:
AMONG US

Kapu1178 committed:
SSzas refactor + initial gas stuff

Kapu1178 committed:
Space wind tweaks

francinum committed:
Merge branch 'master' of github.com:pariahstation/Pariah-Station into zastake2

# Conflicts:
#	_maps/RandomRuins/SpaceRuins/hilbertshoteltestingsite.dmm
#	_maps/RandomZLevels/challenge.dmm
#	_maps/RandomZLevels/wildwest.dmm
#	code/game/objects/effects/effect_system/effects_foam.dm

francinum committed:
Initial Gas Mixture fixes, Admin Debug Verbs

francinum committed:
:screm:

Kapu1178 committed:
Firedoors should no longer cause zone rebuilds

Kapu1178 committed:
Misc fixes

Kapu1178 committed:
Updates c_airblock to check QDELETED

Kapu1178 committed:
Fire stuff - spreading is scuffed

Kapu1178 committed:
FIRE WOOORKSSSS

Kapu1178 committed:
Returns ZAS wait to 2 seconds (sadness)

Kapu1178 committed:
Updates SSzas and Fire

Kapu1178 committed:
Partial revert of MC changes

Kapu1178 committed:
SSairmachines

Kapu1178 committed:
Some fixes. Atmos Sensitive still needs signals

Kapu1178 committed:
Atmos Alarms now work properly & trigger firedoors

Kapu1178 committed:
Attempts to make scrubbers work right, failed

Kapu1178 committed:
Updates SSzas

Kapu1178 committed:
Scrubbers and the beginning of atmos sensitivity removal

Kapu1178 committed:
fixes the overlay code w/ placeholders

Kapu1178 committed:
SSairmachines now recovers during MC failure

Kapu1178 committed:
BURN BABY BURN. And glass cracking sounds

Kapu1178 committed:
Merge branch 'master' of github.com:pariahstation/Pariah-Station into zastake2

Kapu1178 committed:
temp removals

Kapu1178 committed:
All linters fixed

Kapu1178 committed:
grumble

Kapu1178 committed:
grumble 2

Kapu1178 committed:
fuck

Kapu1178 committed:
FUCK

Kapu1178 committed:
Globals

Kapu1178 committed:
wohops

francinum committed:
MINTEST: Atmos Equipment Addition

francinum committed:
Further map updates, add maps.txt def

Kapu1178 committed:
Typo + fixes? passive gates

Kapu1178 committed:
Fixes passive gate for real and fixes a canister bug

Kapu1178 committed:
Hahaha im a dumbass.

francinum committed:
FUCK

francinum committed:
Merge branch 'zastake2' of github.com:Kapu1178/Pariah-Station into zastake2

Kapu1178 committed:
Moar gases, gaseous reagents, and handling.

Kapu1178 committed:
Merge branch 'master' of github.com:pariahstation/Pariah-Station into zastake2

Kapu1178 committed:
Updates fire code to be more functional

Kapu1178 committed:
Removes atmos sensitivity

Kapu1178 committed:
LAVALAND TEST + DEBUG IMPROVEMENT

Kapu1178 committed:
changes for test

francinum committed:
the station is no longer covered in shrek cum

Kapu1178 committed:
lung runtime fix maybe

Kapu1178 committed:
Merge branch 'master' into zastake2

Kapu1178 committed:
Improves airflow behavior significantly

Kapu1178 committed:
Dense objects can now step into non-dense objects during Airflow

Kapu1178 committed:
Micro opt

Kapu1178 committed:
makes all gas mixture procs camelcase (I will regret this later)

Kapu1178 committed:
Clean up, fixes reconcile_air(), recode copyFrom()

Kapu1178 committed:
Fixes critical math errors.

Kapu1178 committed:
Merge branch 'master' of github.com:pariahstation/Pariah-Station into zastake2

francinum committed:
More equipment to test.

Kapu1178 committed:
Fixes gas anal.

Kapu1178 committed:
Fix atmos filters. Quantize gas during filtering

Kapu1178 committed:
Slightly better fix

Kapu1178 committed:
Fixes an error in verb permission.

Kapu1178 committed:
TYPO

Kapu1178 committed:
Bring back pre-init messages

Kapu1178 committed:
Fix some bugs

Kapu1178 committed:
HAHAHAHAHAHHA WHOOPS

Kapu1178 committed:
Fixes kilo. The great unspace-ening

Kapu1178 committed:
Third times the charm

Kapu1178 committed:
Fixes atmos resin.

Kapu1178 committed:
Fixes the portable air pump. I hope.

Kapu1178 committed:
Grammar

Kapu1178 committed:
stupid language

Kapu1178 committed:
Makes some gases purchasable. Oops.

Kapu1178 committed:
Purchaseable air canisters. I hope.

Kapu1178 committed:
Fixes roundstart plasma

Kapu1178 committed:
Canisters have the correct heat resistence

Kapu1178 committed:
Fixes? directional windows blocking atmos due to being constructed on unsimulated turfs

Kapu1178 committed:
Fixes N2O not knocking you the fuck out

Kapu1178 committed:
Fixes atmos consoles and n2o overlays

Kapu1178 committed:
Rewrites vent pumps

Kapu1178 committed:
Tweaks, mainly to airflow

Kapu1178 committed:
Hibernation, filters, optimizations.

Kapu1178 committed:
Merge branch 'master' of github.com:pariahstation/Pariah-Station into zastake2

Kapu1178 committed:
Disables ZASDBG and adds last process

Kapu1178 committed:
Unfucks varedits, optimizes updateValues() by making it a macro

Kapu1178 committed:
lets see if this works i guess

Kapu1178 committed:
Unit Test BS

Kapu1178 committed:
linter appeasement

Kapu1178 committed:
TGM format

Kapu1178 committed:
Whoops

Kapu1178 committed:
brings back precision clean up

Kapu1178 committed:
Pushing this so dreammaker doesnt eat my ass

Kapu1178 committed:
more R-UST work

Kapu1178 committed:
R-UST progress

Kapu1178 committed:
Mo shit

Kapu1178 committed:
more fuel rod stuff

Kapu1178 committed:
softdel my rod

Kapu1178 committed:
grumble

francinum committed:
Merge branch 'master' of github.com:pariahstation/Pariah-Station into pr/Kapu1178/469

Kapu1178 committed:
This is probably fine.

Kapu1178 committed:
Merge branch 'zastake2' of github.com:Kapu1178/Pariah-Station into zastake2

Kapu1178 committed:
Alot of progress. Still fighting the selfmerge bug

Kapu1178 committed:
ZAS READABILITY UPDATE

Adds a ton of comments to ZAS procs, renames some, and greatly improves ZAS debugging tools.

Kapu1178 committed:
REALLY hacky fix to shuttle movement

Kapu1178 committed:
Compile stuff + misleading proc name change

francinum committed:
multizas testing map

Kapu1178 committed:
Enables MultiZAS

Kapu1178 committed:
Try to fix world bounds issue with ZAS

Kapu1178 committed:
Runtimes fix

Kapu1178 committed:
Fixes a hilarious oversight

Kapu1178 committed:
CI

Kapu1178 committed:
please

Kapu1178 committed:
fix icebox

Kapu1178 committed:
Linter pass attempt 1

Kapu1178 committed:
more unused stuff cleanup

Kapu1178 committed:
gas id update

Kapu1178 committed:
more unused stuff

Kapu1178 committed:
mooooore shiit

Kapu1178 committed:
are you fucking kidding me

Kapu1178 committed:
please dont make me force push a reset

Kapu1178 committed:
hard del fix + transfer moles debugging

Kapu1178 committed:
typo

Kapu1178 committed:
try to fix reactor hard del

Kapu1178 committed:
Fix some bugs. fuel_objs no longer leave refs

Kapu1178 committed:
I Am An Idiot (fire compile fix)

Kapu1178 committed:
Fixes mapping error causing a bad init

Kapu1178 committed:
docs zas compile options

Kapu1178 committed:
Disables SSairflow during unit tests.

Tokorizo commented:
My man finnaly did it... all infinity checks in his grasp... so zased.

Kapu1178 committed:
Fixes a critical bug, some more bugs, and comments

Kapu1178 committed:
Disables ZASDBG

Kapu1178 committed:
more comments + another attempt to find this bug

Kapu1178 committed:
fixes an erroneous return value

Kapu1178 committed:
Prevents air from moving up/down in hyperspace

LemonInTheDark commented on code/modules/atmospherics/ZAS/ConnectionGroup.dm:
```diff
@@ -0,0 +1,307 @@
+		var/atom/movable/M = movable[i]
+
+		//If they're already being tossed, don't do it again.
+		if(M.last_airflow > world.time - zas_settings.airflow_delay) continue
+		if(M.airflow_speed) continue
```
hhh

Kapu1178 replied:
Soul

LemonInTheDark commented on code/modules/atmospherics/ZAS/ConnectionGroup.dm:
```diff
@@ -0,0 +1,307 @@
+
+		//Check for knocking people over
+		if(ismob(M) && differential > zas_settings.airflow_stun_pressure)
+			if(M:status_flags & GODMODE) continue
```
weird place to use :

LemonInTheDark commented on code/modules/atmospherics/ZAS/ConnectionGroup.dm:
```diff
@@ -0,0 +1,307 @@
+			if(M:status_flags & GODMODE) continue
+			M:airflow_stun()
+
+		if(M.check_airflow_movable(differential))
```
early continue?

Kapu1178 replied:
good idea

LemonInTheDark commented on code/modules/atmospherics/ZAS/ConnectionGroup.dm:
```diff
@@ -0,0 +1,307 @@
+		if(M.check_airflow_movable(differential))
+			//Check for things that are in range of the midpoint turfs.
+			var/list/close_turfs = list()
+			for(var/turf/U in connecting_turfs)
```
could be cached locally

LemonInTheDark commented on code/modules/atmospherics/ZAS/ConnectionGroup.dm:
```diff
@@ -0,0 +1,307 @@
+
+///Airflow proc causing all objects in movable to be checked against a pressure differential. See file header for more info.
+/connection_edge/proc/flow(list/movable, differential, repelled)
+	for(var/i in 1 to length(movable))
+		var/atom/movable/M = movable[i]
```
why this format for the iteration?

Kapu1178 replied:
Good question. Next question.

LemonInTheDark replied:
no like I think it's straight up just slower here, and I'm not sure what the upside is

Kapu1178 replied:
I was jooooking. I planned on fixing it but got swiftly distracted.

LemonInTheDark commented on code/modules/atmospherics/ZAS/ConnectionGroup.dm:
```diff
@@ -0,0 +1,307 @@
+		var/atom/movable/M = movable[i]
+
+		//If they're already being tossed, don't do it again.
+		if(M.last_airflow > world.time - zas_settings.airflow_delay) continue
```
if it's hot should cache the datum var access (could cache the subtraction really)

LemonInTheDark commented on code/modules/atmospherics/ZAS/ConnectionGroup.dm:
```diff
@@ -0,0 +1,307 @@
+		if(M.airflow_speed) continue
+
+		//Check for knocking people over
+		if(ismob(M) && differential > zas_settings.airflow_stun_pressure)
```
also could cache half this check too

LemonInTheDark commented on code/modules/atmospherics/ZAS/ConnectionGroup.dm:
```diff
@@ -0,0 +1,307 @@
+			//Check for things that are in range of the midpoint turfs.
+			var/list/close_turfs = list()
+			for(var/turf/U in connecting_turfs)
+				if(get_dist(M,U) < world.view) close_turfs += U
+			if(!close_turfs.len) continue
```
:sadkirby:

Kapu1178 replied:
Hehehehe

LemonInTheDark commented on code/modules/atmospherics/ZAS/ConnectionGroup.dm:
```diff
@@ -0,0 +1,307 @@
+			M.airflow_dest = pick(close_turfs) //Pick a random midpoint to fly towards.
+
+			//Soul code im too scared to touch
+			if(repelled) spawn if(M) M.RepelAirflowDest(differential/5)
+			else spawn if(M) M.GotoAirflowDest(differential/10)
```
I am willing to bet these spawns existed to allow for sleep looping, but we have subsystems now

Kapu1178 replied:
I am very scared to change this, unironically.

LemonInTheDark replied:
you know what spawn does, right?

LemonInTheDark reviewed (commented):
pictured: lemon malds about 
`if(case) result`

Kapu1178 committed:
Please don't break anything. Adds the Supermatter

Kapu1178 committed:
Hibernating scrubbers now appear as off

Kapu1178 committed:
Merge branch 'master' of github.com:pariahstation/Pariah-Station into zastake2

Kapu1178 committed:
Whoops. Forgot to stage the N2 tank changes.

Kapu1178 committed:
fix maps

Kapu1178 committed:
re-adds SM monitor

Kapu1178 committed:
Fixes scrubber runtime

Kapu1178 committed:
Hopefully fixes magboots

Kapu1178 committed:
Fixes

Kapu1178 committed:
EXPERIMENTAL: Smart Vents/Pipes

Kapu1178 committed:
Airflow tweaks and improvements

Kapu1178 committed:
TTVs should now function more

Kapu1178 committed:
writing

Kapu1178 committed:
Fixes fire runtime

Kapu1178 committed:
optimizations, gasgas bad

Kapu1178 committed:
Merge branch 'master' of github.com:pariahstation/Pariah-Station into zastake2

Kapu1178 committed:
Liquid plasma now has the proper atmos temp

Kapu1178 committed:
fixes the implosion thingy

Kapu1178 committed:
Fixes is_safe_turf

Kapu1178 committed:
fixes is_safe_turf for real

Kapu1178 committed:
misc tweaks

Kapu1178 committed:
vent pump examine

Kapu1178 committed:
Merge branch 'master' of github.com:pariahstation/Pariah-Station into zastake2

Kapu1178 committed:
HOW DID THIS FIX IT? WHAT? THE FUCK ?????

Kapu1178 committed:
Fuck it. This is fine.

Kapu1178 committed:
oops

Kapu1178 committed:
please dont break anything thx

Kapu1178 committed:
disables debug

francinum committed:
Remove multiline comments in the DME

Kapu1178 committed:
runtime fix

Kapu1178 committed:
Tries to fix knockdown, and buffs extinguisher cooling

Kapu1178 committed:
help me debug pls

Kapu1178 committed:
"air" is now a valid label for canisers

Kapu1178 committed:
filter changes

Kapu1178 committed:
Fixes pluox and tritium :)

Kapu1178 committed:
air alarms now have the full gas list + reordered

Kapu1178 committed:
Re-orders common gas list

Kapu1178 committed:
Makes deuterium isotope 2

Kapu1178 committed:
vent and scrubber controls now reset hibernation

Kapu1178 committed:
Fixes portable scrubbers, changes to air alarms

Kapu1178 committed:
Explosion pass

Kapu1178 committed:
Fixes

Kapu1178 committed:
Fixes atmozphere program. fuck js

Kapu1178 committed:
Reworks some lists n' things

Kapu1178 committed:
removes redundant global list

Kapu1178 committed:
Back-to-basics firelocks, pt1

Kapu1178 committed:
fire/air alarm + firedoor refactor. Fuck me harder

Kapu1178 committed:
IM A DUMBASS IM A DUMBASS IM A DUMBASS

Kapu1178 committed:
ashlizard lungs

Kapu1178 committed:
fix compile

Kapu1178 committed:
Adds airalarm firealerts and thermostats

francinum reviewed (approved):
LGTM

Kapu1178 committed:
fix some runtimes

Kapu1178 committed:
oops

Kapu1178 committed:
I CLICKED STAGE ALL FOR FUCK SAKE VSC

Kapu1178 committed:
Fix fire

Kapu1178 committed:
Merge branch 'master' into zastake2

Kapu1178 committed:
Merge branch 'master' of github.com:pariahstation/Pariah-Station into zastake2

Kapu1178 committed:
removes leftover testing artifact

Kapu1178 committed:
for real this time

Kapu1178 committed:
Removes unused files

Kapu1178 committed:
Fixes nitrodioxide label

Kapu1178 committed:
fixes methyl bromide label

Kapu1178 committed:
micro opts

Kapu1178 committed:
Tramstation should no longer freeze the server

Kapu1178 committed:
Removes minimum zone size

Kapu1178 committed:
Linterinos

Kapu1178 committed:
linterinos for real

Kapu1178 committed:
Reverts zone size change, makes other tweaks

Kapu1178 committed:
Tries to fix the phantom rest bug

Kapu1178 committed:
airflow fix
